### ISSUE TITLE
Performance Benchmark?

### ISSUE THREAD

**Issue by The-Don-Himself** [varnish/varnish-devicedetect#28]
Greetings,

I really want to try this out. I have a <a href="https://campus-discounts.com" target="_blank">Symfony app</a> running behind Varnish and the performance is really good, but as always, it can be be better.

All requests must reach the app then, depending on certain parameters, a specific response is generated. I cache db heavy fragments of the page using ESI. One of those parameters is Device detection. I currently use https://github.com/serbanghita/Mobile-Detect.

I have been profiling my app and the response times for the home page is 179ms, a whooping 125ms of which is for device detection.

I am curious to the performance of mobile detection using Varnish, has anyone ever benchmarked it out? Of course, written in C should make it really fast but how much? I know the best way is to profile myself but you see, it will mean changing my app code entirely which will take days, I was wondering if anyone had rough estimates/ideas.

Regards.

**Comment by quintonparker** [comment]
We run a large-ish ecommerce store and we use varnish devicedetect for millions of requests per month. We're able to serve the full request (include a devicedetect call) from varnish in as little as 1e-4 ie. less than a millisecond

i guess this is in part due to pcre jit which should be enabled by default in most varnish installs

**Comment by The-Don-Himself** [comment]
@quintonparker 
Wow, thank you for the prompt response!

So in your option, using this script to detect a device will not take 125ms, in fact, in your experience it should take less than 1ms!? This is awesome, I must try it and although I will close this issue, I will update it with any benchmarks I get from my app with mobile detect in place.

Cheers!

**Comment by quintonparker** [comment]
You're welcome. You can even apply an optimisation on top of that by instructing browsers/agents to save the result as a cookie and so you can bypass the device detection and use the value from the cookie instead

**Comment by The-Don-Himself** [comment]
@quintonparker 
By the way, I just realized that to benchmark it, I do not have to change my application code. I just installed it on varnish and checked if its working by manipulating the browser agent in Chrome console. After confirming that it was working, I compared the load times before and after installation. Your 1ms (at worst) seems spot on!

To integrate this, I must however, change alot in my app but make no mistake, it already seems totally worth it! Thanks.
